mountpoint-s3/src/inode.rs

```diff
@@ -360,6 +360,68 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Unlink the entry described by `parent_ino` and `name`.
+    ///
+    /// If the entry exists, delete it from S3 and the superblock.
+    pub async fn unlink<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        parent_ino: InodeNo,
+        name: &OsStr,
+    ) -> Result<(), InodeError> {
+        let parent = self.inner.get(parent_ino)?;
+        let LookedUp { inode, .. } = self.lookup(client, parent_ino, name).await?;
+
+        if inode.kind() == InodeKind::Directory {
+            return Err(InodeError::IsDirectory(inode.ino()));
+        }
+
+        let write_status = {
+            let inode_state = inode.inner.sync.read().unwrap();
+            inode_state.write_status
+        };
+
+        match write_status {
+            WriteStatus::LocalUnopened | WriteStatus::LocalOpen => {
+                // In the future, we may permit `unlink` and cancel any in-flight uploads.
+                debug!("unlink called on local file, unlink not supported until write is complete");
+                return Err(InodeError::UnlinkNotPermittedWhileWriting(inode.ino()));
+            }
+            WriteStatus::Remote => {
+                debug!("unlink called on remote file, will delete from S3 and unlink from parent");
+                let (bucket, s3_key) = (self.inner.bucket.as_str(), inode.full_key());
+                let delete_obj_result = client.delete_object(bucket, s3_key).await;
+
+                match delete_obj_result {
+                    Ok(_res) => debug!("object with key {s3_key} successfully deleted, if it existed"),
+                    Err(e) => {
+                        error!(error=?e, "unlink failed when trying to perform S3 DeleteObject call, not unlinking from parent inode");
+                        Err(InodeError::ClientError(e.into()))?;
+                    }
+                };
+            }
+        }
+
+        let mut parent_state = parent.inner.sync.write().unwrap();
+        match &mut parent_state.kind_data {
+            InodeKindData::File { .. } => {
+                debug_assert!(false, "inodes never change kind");
+                return Err(InodeError::NotADirectory(parent.ino()));
+            }
+            InodeKindData::Directory {
+                children,
+                writing_children,
+            } => {
+                writing_children.remove(&inode.ino());
```

Comment:
We can probably check `if writing_children.contains(&inode.ino())`, and if there is new local file with same name being created, we may abort our operation, as new local file will overwrite the remote object anyway?

Reply:
Agree that would be the best solution if we don't change how we lock / serialize operations for unlink.

Based on James' feedback, I'm thinking that tracking 'delete status' is the way to move forward.

Reply:
I think we need to tackle the concurrency of `unlink` head on - this might be a solution, or maybe there's an alternative. Let's discuss in the other thread (https://github.com/awslabs/mountpoint-s3/pull/232/files#r1186043700).